packages/special-pages/pages/new-tab/app/privacy-stats/PrivacyStatsProvider.js

```diff
@@ -0,0 +1,245 @@
+        default:
+            return state
+        }
+    }
+    default:
+        return state
+    }
+}
+
+/**
+ * @param {object} params
+ * @param {import("preact/hooks").Dispatch<Events>} params.dispatch
+ * @param {import("preact").RefObject<PrivacyStatsService>} params.service
+ */
+function useInitialData ({ dispatch, service }) {
+    useEffect(() => {
+        if (!service.current) return console.warn('missing service')
+        const stats = service.current
+        async function init () {
+            const { config, data } = await stats.getInitial()
+            if (data) {
+                dispatch({ kind: 'initial-data', data, config })
+            } else {
+                dispatch({ kind: 'error', error: 'missing data from getPrivacyStats' })
+            }
+        }
+
+        dispatch({ kind: 'load-initial' })
+
+        // eslint-disable-next-line promise/prefer-await-to-then
+        init().catch((e) => {
+            console.error('uncaught error', e)
+            dispatch({ kind: 'error', error: e })
+        })
+
+        return () => {
+            stats.destroy()
+        }
+    }, [])
+}
+
+/**
+ * @return {import("preact").RefObject<PrivacyStatsService>}
+ */
+function useService () {
+    const service = useRef(/** @type {PrivacyStatsService|null} */(null))
+    const ntp = useMessaging()
+    useEffect(() => {
+        const stats = new PrivacyStatsService(ntp)
+        service.current = stats
+        return () => {
+            stats.destroy()
+        }
+    }, [ntp])
+    return service
+}
+
+/**
+ * Subscribe to data updates
+ * @param {object} params
+ * @param {import("preact/hooks").Dispatch<Events>} params.dispatch
+ * @param {import("preact").RefObject<PrivacyStatsService>} params.service
+ */
+function useDataSubscription ({ dispatch, service }) {
+    useEffect(() => {
+        if (!service.current) return console.warn('could not access service')
+        const unsub = service.current.onData((evt) => {
+            dispatch({ kind: 'data', data: evt.data })
+        })
+        return () => {
+            unsub()
+        }
+    }, [service, dispatch])
+}
+
+/**
+ * Subscribe to config updates + provide a method for toggling visibility
+ * @param {object} params
+ * @param {import("preact/hooks").Dispatch<Events>} params.dispatch
+ * @param {import("preact").RefObject<PrivacyStatsService>} params.service
+ */
+function useConfigSubscription ({ dispatch, service }) {
+    const { isReducedMotion } = useEnv()
+
+    /**
+     * Consumers can call 'toggle' and the in-memory data will be updated in the service
+     * The result of that toggle will be broadcasts immediately, allowing real-time (optimistic) UI updates
+     */
+    const toggle = useCallback(() => {
+        service.current?.toggleExpansion()
+    }, [service, dispatch])
+
+    useEffect(() => {
+        if (!service.current) return console.warn('could not access service')
+        const unsub2 = service.current.onConfig((data) => {
```

Comment:
Is unsub2 named that way just to differentiate it from the Data unsub function? Or is it to avoid a scoping issue I’m not seeing?

Reply:
yeah, there was previously 2 items here, I changed it